src/main/java/frc/robot/Manipulator.java

```diff
@@ -266,38 +268,117 @@ public void startLogging(String basekey) {
+	}
+	public ManipulatorControl controlManipulator(
+		DoubleSupplier a, DoubleSupplier g, DoubleSupplier w,
+		BooleanSupplier wr, BooleanSupplier al, BooleanSupplier gl
+	) {
+		return new ManipulatorControl(this, a, g, w, wr, al, gl);
+	}
+
+
 
 
 
-	public static class TestManipulator extends CommandBase {
+	public static class ManipulatorControl extends CommandBase {
 
 		public static final double
 			ARM_WINCH_VOLTAGE_SCALE = 5.0,
-			GRAB_CLAW_VOLTAGE_SCALE = 7.0;
+			ARM_WINCH_LOCK_VOLTAGE = 2.5,
+			GRAB_CLAW_VOLTAGE_SCALE = 7.0,
+			GRAB_CLAW_LOCK_VOLTAGE = 2.5,
+			WRIST_ACCUMULATION_RATE_SCALE = 0.01;	// at full throttle, add 0.01 x 50 loops per second = 0.5 per second change [maximum]
 
 		private final Manipulator
 			manipulator;
 		private final DoubleSupplier
-			arm_percent,
-			grab_percent,
-			wrist_set;
-
-		public TestManipulator(Manipulator m, DoubleSupplier a, DoubleSupplier g, DoubleSupplier w) {
+			arm_rate,
+			grab_rate,
+			wrist_rate;
+		private final BooleanSupplier
+			wrist_reset,
+			arm_lock,
+			grab_lock;
+
+		private double
+			wrist_position = Grabber.WRIST_NEUTRAL_PERCENT;
+		private boolean
+			is_arm_locked = false,
+			is_grab_locked = false;
+
+
+		public ManipulatorControl(
+			Manipulator m,
+			DoubleSupplier a, DoubleSupplier g, DoubleSupplier w
+		) {
 			this.manipulator = m;
-			this.arm_percent = a;
-			this.grab_percent = g;
-			this.wrist_set = w;
-		}
+			this.arm_rate = a;
+			this.grab_rate = g;
+			this.wrist_rate = w;
+			this.wrist_reset = this.arm_lock = this.grab_lock = ()->false;
+			super.addRequirements(m.arm, m.grabber);
+		}
+		public ManipulatorControl(
+			Manipulator m,
+			DoubleSupplier a, DoubleSupplier g, DoubleSupplier w,
+			BooleanSupplier wr, BooleanSupplier al, BooleanSupplier gl
+		) {
+			this.manipulator = m;
+			this.arm_rate = a;
+			this.grab_rate = g;
+			this.wrist_rate = w;
+			this.wrist_reset = wr;
+			this.arm_lock = al;
+			this.grab_lock = gl;
+			super.addRequirements(m.arm, m.grabber);
+		}
+
+
+		// private double calcWristPos()
+		// {
+		// 	if(this.wrist_set.getAsDouble() > 0 && wristPosition < 1.0)
+		// 	{
+		// 		return wrist_position += 0.1;
+		// 	}
+		// 	else if(this.wrist_set.getAsDouble() < 0 && wristPosition > 0.0)
+		// 	{
+		// 		return wrist_position -= 0.1;
+		// 	}
+		// 	else 
+		// 	{
+		// 		return wrist_position;
+		// 	}
+		// }
 
 		@Override
 		public void initialize() {
-			// System.out.println("Starting manipulator test!");
+			this.wrist_position = Grabber.WRIST_NEUTRAL_PERCENT;
 		}
 		@Override
 		public void execute() {
-			this.manipulator.arm.setWinchVoltage(this.arm_percent.getAsDouble() * ARM_WINCH_VOLTAGE_SCALE);
-			this.manipulator.grabber.setGrabberVoltage(this.grab_percent.getAsDouble() * GRAB_CLAW_VOLTAGE_SCALE);
-			this.manipulator.grabber.setWristPercent(this.wrist_set.getAsDouble());
+			this.wrist_position += this.wrist_rate.getAsDouble() * WRIST_ACCUMULATION_RATE_SCALE;
```

Comment:
This looks like it is new code that uses the position of the XBox control stick to set the wrist velocity instead of the wrist position, correct?  Sam and Zoke said this code wasn't quite working right.  Can we fall back to the old way if necessary if we can't get it working the way we want by tomorrow?

Reply:
This is my implementation which was working but was in the wrong direction. I already fixed that so it should be fine.